src/Inciter/AMR/node_connectivity.h

```diff
@@ -0,0 +1,165 @@
+             */
+            static size_t face_list_opposite(face_list_t face_list, size_t opposite_index)
+            {
+                // TODO: make this actually inspect the face_list and be much
+                // more robust...
+                // TODO: Remove this hack to supress warning
+                size_t result = face_list[0][0];
+                switch(opposite_index)
+                {
+                    case 0:  // ABC
+                        result = 3;
+                        break;
+                    case 1:  // ABD
+                        result = 2;
+                        break;
+                    case 2:  // ACD
+                        result = 1;
+                        break;
+                    case 3:  // BCD
+                        result = 0;
+                        break;
+                    default: // something went horribly wrong..
+                        assert(0);
+                        break;
+                }
+
+                return result;
+            }
+
+            // TODO: Document this
+            // Int because it's signed.. is this a good idea?
+            int find(size_t A, size_t B)
+            {
+                size_t min = std::min(A,B);
+                size_t max = std::max(A,B);
+
+                for (size_t i = 0; i < size(); i++)
+                {
+                    // Did we find it?
+                    node_pair_t n = get(i);
+                    if (min == n[0] && max == n[1])
+                    {
+                        return static_cast<int>(i);
+                    }
+                }
+                return -1;
+            }
+
+            // TODO: Document this
+            size_t add(size_t A, size_t B)
+            {
+                if (A != 0 || B != 0)
+                {
+                    assert(A != B);
+                    // TODO: Abstract to exists method. (Could have one for id,
+                    // as well as one for val)
+
+                    // check if already exists
+                    int f = find(A,B);
+                    if (f != -1) {
+                        std::cout << "Connect already exits " << A << " " << B << std::endl;
+                        return static_cast<size_t>(f);
+                    }
+                }
+
+                nodes.push_back( {{std::min(A,B), std::max(A,B)}} );
+                std::cout << " add " << size()-1 << " a = " <<  A << " b = " << B << std::endl;
+                return size()-1;
+            }
+
+            /**
+             * @brief Print connectivity a id: a-b
+             */
+            void print()
+            {
+                std::cout << "Connectivity" << std::endl;
+                for (size_t i = 0; i < size(); i ++)
+                {
+                    std::cout << i << ": A " << get(i)[0] << " B " << get(i)[1] << std::endl;
+                }
+            }
+
+    };
+}
+
+#endif // guard
```

Comment:
Yeah. `std::vector<T>` itself is guaranteed to be contiguously storing `T`. But if `T` is a container with its own heap-based storage, it will not be contiguous anymore, e.g., a vec of vecs. Having said that a vec of arrays might be. I'm actually not sure. But converting to just a vec of Ts will definitely be contiguous.

This is actually, not a big deal at all at this point; just wanted point out, so that we think about and perhaps put it on a todo list. Most important at this point is to put merge this in if it is correct. We can continue cleaning this up in following passes.